Allow three extra refinement levels for an ElementId

kidder opened the pull request:
This is done by using a compact representation of a SegmentId. Also fix some clang-tidy errors.

## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder committed:
Allow three extra refinement levels for an ElementId

This is done by using a compact representation of a SegmentId.
Also fix some clang-tidy errors.

nilsdeppe commented on src/Utilities/StdHelpers/Bit.hpp:
```diff
@@ -0,0 +1,34 @@
+constexpr bool has_single_bit(T x) noexcept {
+  return std::popcount(x) == 1;
+}
+}  // namespace std
```
We should only be adding this into the std namespace if it's not defined in the standard, otherwise we hit UB. There's is identically 0 chance that people will correctly place the if-else checks for the includes in all the right places.

Also, please make adding this a separate commit.

Make sure to add a code comment into this file explaining why this was necessary.

nilsdeppe commented on src/Domain/Structure/SegmentId.cpp:
```diff
@@ -12,7 +14,7 @@
 SegmentId::SegmentId(const size_t refinement_level, const size_t index)
     : refinement_level_(refinement_level), index_(index) {
-  ASSERT(refinement_level < max_refinement_level,
+  ASSERT(refinement_level <= max_refinement_level,
```
Just remove this check. It's not a constraint on `SegmentId` but a constraint on `ElementId`

nilsdeppe commented on src/Domain/Structure/ElementId.cpp:
```diff
@@ -3,29 +3,37 @@
+#include <bit>
+#else
+#include "Utilities/StdHelpers/Bit.hpp"
+#endif
```
This check should be in the `Bit.hpp` file. Nobody will know to do this magic when including that file.

nilsdeppe commented on src/Domain/Structure/ElementId.cpp:
```diff
@@ -40,17 +48,88 @@ static_assert(sizeof(ElementId<2>) == 2 * sizeof(int),
+static_assert(ElementId<2>::max_refinement_level ==
+              SegmentId::max_refinement_level);
+static_assert(ElementId<3>::max_refinement_level ==
+              SegmentId::max_refinement_level);
```
I think we can remove these checks since there's no actual restriction on `SegmentId`

nilsdeppe commented on src/Domain/Structure/ElementId.cpp:
```diff
@@ -40,17 +48,88 @@ static_assert(sizeof(ElementId<2>) == 2 * sizeof(int),
+}
+
+SegmentId make_segment_id(const uint16_t compact_segment_id) {
+  return SegmentId{get_refinement_level(compact_segment_id),
```
Add the `ASSERT` here too about  undefined compact segment id?

nilsdeppe commented on src/Domain/Structure/ElementId.cpp:
```diff
@@ -40,17 +48,88 @@ static_assert(sizeof(ElementId<2>) == 2 * sizeof(int),
+         static_cast<size_t>(std::countl_zero(compact_segment_id));
+}
+
+size_t get_index(uint16_t compact_segment_id) {
```
`const uint16_t`?

nilsdeppe commented on src/Domain/Structure/ElementId.cpp:
```diff
@@ -40,17 +48,88 @@ static_assert(sizeof(ElementId<2>) == 2 * sizeof(int),
+         static_cast<uint16_t>(segment_id.index());
+}
+
+size_t get_refinement_level(uint16_t compact_segment_id) {
```
`const uint16_t`?

nilsdeppe commented on src/Domain/Structure/ElementId.cpp:
```diff
@@ -40,17 +48,88 @@ static_assert(sizeof(ElementId<2>) == 2 * sizeof(int),
+bool is_on_upper_block_boundary(const uint16_t compact_segment_id) {
+  ASSERT(compact_segment_id > 0, "Undefined compact segment id");
+  // true if all bits after the leading one are also one
+  return 16 == std::countl_zero(compact_segment_id) +
```
I would find this easier to read as `16 == (...)`

nilsdeppe commented on src/Domain/Structure/ElementId.cpp:
```diff
@@ -40,17 +48,88 @@ static_assert(sizeof(ElementId<2>) == 2 * sizeof(int),
+                                uint8_t direction,
+                                uint16_t compact_segment_id_xi,
+                                uint16_t compact_segment_id_eta,
+                                uint16_t compact_segment_id_zeta)
```
`const uint?_t`

nilsvu reviewed (dismissed):
This is pretty cool 😎 thanks for doing this.

nilsdeppe reviewed (commented):
LGTM, could you move the adding of `std::bit_*` into a separate commit as you squash?

kidder committed:
Add StdHelper for <bit>

Some useful functions (std::has_single_bit and std::bit_floor) in
the standard library bit are not defined until the version used by
gcc10.  So if they are not in the library we add implementations
based on cppreference.com
